puppet/lib/puppet/util/puppetdb/char_encoding.rb

```diff
@@ -206,6 +220,137 @@ def self.get_byte(str, index)
     end
   end
 
+  # This EncodingAnalyser is supposed to be used by this module only it tries
+  # to break the string into smaller pieces in order to make line-wise diffing
+  # possible.
+  #
+  # In most circumstances the diff should give enough context to understand the
+  # issue better.
+  class EncodingAnalyser
+    class String
+      attr_reader :filename
+
+      def initialize(string, base_filename, type)
+        @string   = string
+        @type     = type
+        @filename = "#{base_filename}-#{@type}"
+        @errors   = []
+      end
+
+      def encoding
+        @string.encoding
+      end
+
+      def size
+        @string.size
+      end
+
+      def write
+        File.open(@filename, 'w') do |file|
+          file.flock(File::LOCK_EX)
+          file.puts split
+          file.chmod(0666)
+          file.flock(File::LOCK_UN)
+        end
+        filename
+      end
+
+      def errors
+        if @errors.any?
+          ["Errors in #{@type} string", @errors].join("\n")
+        else
+          ''
+        end
+      end
+
+      private
+
+      # Splitting the string does the following two things:
+      #   - split hashes and array that are values from their keys
+      #   - split lists of hashes into new lines
+      def split(second_try = false)
+        begin
+          @string.gsub(/(['"]:)({|\[)/, "\\1\n\\2").gsub(/(}|\]),({|\[)/, "\\1,\n\\2")
+        rescue ArgumentError => e
+          if second_try
+            @errors << "The #{@type} string cannot be split into diffable chunks."
+            @string
+          else
+            @string.encode!(@string.encoding, :invalid => :replace, :undef => :replace, :replace => "?")
+            @errors << "Invalid bytes encountered. These have been replaced with '?'."
+            @errors << "This should be enough to produce a helpful diff."
+            split(:second_try)
+          end
+        end
+      end
+    end
+
+    def initialize(filename)
+      @filename = filename
+    end
+
+    def original= string
+      @original = String.new(string, @filename, 'original')
+    end
+
+    def converted= string
+      @converted = String.new(string, @filename, 'converted')
+    end
+
+    def write_summary
+      File.open(@filename, 'w') do |file|
+        file.flock(File::LOCK_EX)
+        file.puts diff_header
+
+        orig_fn = @original.write
+        conv_fn = @converted.write
+
+        file.puts "diff of #{orig_fn} and #{conv_fn}"
+        file.puts ""
+        file.puts `diff -u5 #{orig_fn} #{conv_fn}`
```

Comment:
External shell dependencies are probably not a great idea here, this terminus primarily now runs in the JVM given that puppet-server is the way forward for puppet masters. I haven't looked into this nor proven it myself, but I presume that a JRuby exec like this uses the clone system call in linux, which requires initially the same amount of RAM as the original JVM before it shrinks down to the necessary size. This can cause linux OOM_Killer to potentially initiate an OOM kill routine at time, unless administrators have tuned for this eventuation. I'm not a massive fan of putting this kind of risk to our users.

I think, if we really wanted to go down the direction this PR states, perhaps an avoidance of external exec's is a good idea. One can always place the two files to be diffed manually instead. Otherwise a diff that doesn't require an exec would be preferred, if such a thing exists without say - C or external dependencies in Ruby.

Not to mention GNU diff and BSD diff can be different, having something hard-coded like this can create compatibility hassles.

Anyway, you get the idea, shell execs are hard for multiple reasons.